openai-model-provider/proxy/rewrite.go

```diff
@@ -117,3 +117,59 @@ func RewriteAllModelsWithUsage(usage string, filters ...func(string) bool) func(
 		return nil
 	}
 }
+
+// RewriteAllModelsWithUsageMap returns a response modifier that marks all models with the specified usage
+func RewriteAllModelsWithUsageMap(usageMap map[string][]func(string) bool) func(*http.Response) error {
+	return func(resp *http.Response) error {
+		if resp.StatusCode != http.StatusOK {
+			return nil
+		}
+
+		defer resp.Body.Close()
+
+		var body io.Reader = resp.Body
+		if resp.Header.Get("Content-Encoding") == "gzip" {
+			gzReader, err := gzip.NewReader(resp.Body)
+			if err != nil {
+				return fmt.Errorf("failed to create gzip reader: %w", err)
+			}
+			defer gzReader.Close()
+			resp.Header.Del("Content-Encoding")
+			body = gzReader
+		}
+
+		var models api.ModelsResponse
+		if err := json.NewDecoder(body).Decode(&models); err != nil {
+			return fmt.Errorf("failed to decode models response: %w", err)
+		}
+
+		for usage, filters := range usageMap {
+			for i, model := range models.Data {
+				if model.Metadata == nil {
+					model.Metadata = make(map[string]string)
+				}
+
+				if len(filters) == 0 {
+					model.Metadata["usage"] = usage
+				} else {
+					for _, filter := range filters {
+						if filter(model.ID) {
+							model.Metadata["usage"] = usage
+							break
+						}
+					}
+				}
+				models.Data[i] = model
+			}
+		}
```

Comment:
The design of this could technically allow for one model to have the usage set to one thing, and then have it get overwritten by a later iteration if it matches the filters for a different usage. Do we want to do that? Or should we add something like this to the loop to skip models that already have a usage?

```go
if model.Metadata["usage"] != "" {
    continue
}
```

Reply:
For xai, it doesn't have usage. So we are just grouping usage based on its own name.

Reply:
Yeah I was just thinking for the case of reusability, since this is in the package that the other model providers depend on

Reply:
Added the skip logic you suggested so that if it has usage then it will skip.